[image_picker] Fix to keep Exif and so on in case of iOS.

ko2ic opened the pull request:
## Description

The image that can be acquired with image_picker is missing some information, even in the case of the original size.

1. ```CreateDate``` of Exif.
2. ```GPSLatitude```,  ```GPSLatitude```, ```GPSLongitudeRef ```, ```GPSLongitude``` of GPS Tags.
3. Animation of GIF
4. PNG transparency (It solves with https://github.com/flutter/plugins/pull/742, but the solution method is different )

This pull request is a modification of these.

Limitation

* The Image of tiff and Unknown magic byte draws as JPEG. (But this has the same problem with the current version.) 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

tvolkert commented:
/cc @xster

collinjackson commented:
@goderbauer I think you were looking at this issue. does it still need merging?

goderbauer commented:
@collinjackson I looked at (and fixed) EXIF data for the camera plugin. The image_picker uses a different API and I wouldn't be surprised if it has the same issue. So, this PR is probably still relevant.

9thwall commented:
I'm running image_picker 0.4.10 on a Pixel 3 phone and the image is rotating as everyone has mentioned. 
Is there a fix we can apply to resolve this?

manhnhhdvn commented:
@collinjackson, @goderbauer: I need this feature, so I fetched this PR then gave it a try, and everything worked fine for me. If this PR has no problems, could you merge it asap?

ko2ic commented:
I checked the operation after a long time, so I  fixed it because there was a bug.

* Fixed was not working with heic.
* Fixed not working with first authentication permission

Are there any plans for this Pull Request to be merged?

ko2ic commented:
@collinjackson This PullRequest corresponds at least to png transparency on ios.
However, merge is necessary for https://github.com/flutter/plugins/pull/742 .
This is difficult to merge. Because in my PR, I use the first line of byte to judge whether it is PNG or not.( https://github.com/flutter/plugins/pull/742 use alpha info.)

If my judgment is okay, Could ios overwrite it with my source?

In addition, this pull request is useful as at least creationDate and GPS information is hold.

Could you make a review because I want to use exif by any means?

romaluca commented:
Up

cyanglaz commented on packages/image_picker/ios/Classes/ImagePickerPlugin.m:
```diff
@@ -150,6 +166,11 @@ - (void)imagePickerController:(UIImagePickerController *)picker
     }
   } else {
+    NSURL *assetURL = [info objectForKey:UIImagePickerControllerReferenceURL];
+    PHFetchResult<PHAsset *> *result = [PHAsset fetchAssetsWithALAssetURLs:@[ assetURL ]
```
`assetURL` can be nil if user take a fresh picture in the image_picker, which leads to crash here.

cyanglaz commented on packages/image_picker/ios/Classes/ImagePickerPlugin.m:
```diff
@@ -254,4 +319,94 @@ - (UIImage *)scaledImage:(UIImage *)image
+}
+
+- (void)createImageFileAtPath:(NSString *)path contents:(NSData *)data with:(PHAsset *)asset {
+  NSMutableDictionary *exifDict = [self fetchExifFrom:asset];
```
Manually parsing the meta data seems to be risky. Is it possible to get the meta data from the original image directly and pass down to the new image? 
Maybe when fetching image in `requestImageDataForAsset`, we can call `(NSDictionary *)CFBridgingRelease(CGImageSourceCopyPropertiesAtIndex(cgImage,0,NULL));` to get the whole meta data of the old image, and then somehow get a diff between the new image and the old image, then add the necessary stuff to the new image?
This way we don't manually parse the information like formatting time, adding hard coded GPS keys etc.

cyanglaz commented on packages/image_picker/ios/Classes/ImagePickerPlugin.m:
```diff
@@ -254,4 +319,94 @@ - (UIImage *)scaledImage:(UIImage *)image
+  _arguments = nil;
+}
+
+- (NSMutableDictionary *)fetchExifFrom:(PHAsset *)asset {
```
If a manual parser is inevitable, let's put the meta data handling into a separate file/class.

cyanglaz commented:
@ko2ic Are you able to keep working on this PR? I think this is valuable. If you are not able to continue, do you mind me taking this over?

ko2ic commented:
@cyanglaz I will hand over to you.
You will probably get better done than I do.

ko2ic commented on packages/image_picker/ios/Classes/ImagePickerPlugin.m:
```diff
@@ -254,4 +319,94 @@ - (UIImage *)scaledImage:(UIImage *)image
   return scaledImage;
 }
 
+- (void)createImageFileAtPath:(NSString *)path contents:(NSData *)data {
```
This mimics the signature of ```NSFileManager``` class.
So I feel good with ```contents```.
However, I leave it to you.

```- (BOOL)createFileAtPath:(NSString *)path contents:(nullable NSData *)data attributes:(nullable NSDictionary<NSFileAttributeKey, id> *)attr;```

cyanglaz commented:
Thanks to @ko2ic for the inspiration provided in this PR. 
The issue this PR tries to fix is fixed after https://github.com/flutter/plugins/pull/1617
The only missing part is gif handling, since it is not clear that what we want to do with scaled GIFs (it seems to lose animation at this point)
I will close this PR now. If anyone has an idea to completely handle GIFs, please feel empowered to open a PR! @ko2ic
